File: my-folder/0064-minimum-path-sum/solution.go
func minPathSum(grid [][]int) int {
    m := len(grid)
    n := len(grid[0])
    memo := make([][]int, m)

    for i := range m {
        memo[i] = make([]int, n)
        for j := range memo[i] {
            memo[i][j] = -1
        }
    }

    var dfs func(int, int) int
    dfs = func(i, j int) int {
        if i >= m || j >= n {
            return math.MaxInt32
        }
        if i == m-1 && j == n-1 {
            return grid[i][j]
        }

        if memo[i][j] != -1 {
            return memo[i][j]
        }

        memo[i][j] = grid[i][j] + min(dfs(i + 1, j), dfs(i, j+1))
        return memo[i][j]
    }

    return dfs(0, 0)
}

File: my-folder/0064-minimum-path-sum/solution.rs
impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let row: usize = grid.len();
        let col: usize = grid[0].len();

        let mut dp: Vec<i32> = vec![i32::MAX; col + 1];
        dp[col-1] = 0;

        for r in (0..row).rev() {
            for c in (0..col).rev() {
                dp[c] = grid[r][c] + dp[c].min(dp[c+1]);
            }
        }

        return dp[0];
    }
}
